src/Cryptol/Transform/Specialize.hs

```diff
@@ -102,6 +104,13 @@ specializeExpr expr =
     EProofAbs p e -> EProofAbs p <$> specializeExpr e
     EProofApp {}  -> specializeConst expr
     EWhere e dgs  -> specializeEWhere e dgs
+    -- The type should be monomorphic, and the guarded expressions should
+    -- already be normalized, so we just need to choose the first expression
+    -- that's true.
+    EPropGuards guards _schema -> 
+      case List.find (all checkProp . fst) guards of
+        Just (_, e) -> specializeExpr e
+        Nothing -> fail "No guard constraint was satisfied"
 
```

Comment:
This probably shouldn't fail, but rather emit an expression that will raise an exception when evaluated (e.g., `error`)

Reply:
I defined an `eError'` that is the same as `eError` but doesn't explicitly apply a type, and used it here